m4/mpi.m4

```diff
@@ -145,31 +145,47 @@ AS_IF([test "x$MPI_IMPL" != x],
         AS_IF([test -e $MPI_INCLUDES_PATH/mpi.h],
               [
                 AS_ECHO(["note: using $MPI_INCLUDES_PATH/mpi.h"])
-                tmpCPPFLAGS=$CPPFLAGS
-                AC_LANG_SAVE
-                AC_LANG_CPLUSPLUS
-                CPPFLAGS="-I$MPI_INCLUDES_PATH $CPPFLAGS"
-                AC_CHECK_HEADER([mpi.h],
-                                [AC_DEFINE(HAVE_MPI, 1, [Flag indicating whether or not MPI is available])],
-                                [AC_MSG_RESULT([Could not compile in the MPI headers...]); enablempi=no])
-                MPI_INCLUDES_PATHS="-I$MPI_INCLUDES_PATH"
-                AC_LANG_RESTORE
-                CPPFLAGS=$tmpCPPFLAGS
+                mpimajor=`grep "OMPI_MAJOR_VERSION" $MPI_INCLUDES_PATH/mpi.h | sed -e "s/#define OMPI_MAJOR_VERSION[ ]*//g"`
+                AS_IF([test "$mpimajor" = "1"],
+                      [
+                     enablempi=no
+                     AC_MSG_RESULT([<<< MPI version 1.x detected, which is not supported  MPI disabled. >>>])
+                     ],
+                     [
+                     tmpCPPFLAGS=$CPPFLAGS
+                     AC_LANG_SAVE
+                     AC_LANG_CPLUSPLUS
+                     CPPFLAGS="-I$MPI_INCLUDES_PATH $CPPFLAGS"
+                     AC_CHECK_HEADER([mpi.h],
+                                     [AC_DEFINE(HAVE_MPI, 1, [Flag indicating whether or not MPI is available])],
+                                     [AC_MSG_RESULT([Could not compile in the MPI headers...]); enablempi=no])
+                     MPI_INCLUDES_PATHS="-I$MPI_INCLUDES_PATH"
+                     AC_LANG_RESTORE
+                     CPPFLAGS=$tmpCPPFLAGS
+                     ])
               ],
               [test -e $MPI_INCLUDES_PATH/mpi/mpi.h],
               [
-                MPI_INCLUDES_PATH=$MPI_INCLUDES_PATH/mpi
-                AS_ECHO(["note: using $MPI_INCLUDES_PATH/mpi.h"])
-                tmpCPPFLAGS=$CPPFLAGS
-                AC_LANG_SAVE
-                AC_LANG_CPLUSPLUS
-                CPPFLAGS="-I$MPI_INCLUDES_PATH $CPPFLAGS"
-                AC_CHECK_HEADER([mpi.h],
-                                [AC_DEFINE(HAVE_MPI, 1, [Flag indicating whether or not MPI is available])],
-                                [AC_MSG_RESULT([Could not compile in the MPI headers...]); enablempi=no] )
-                MPI_INCLUDES_PATHS="-I$MPI_INCLUDES_PATH"
-                AC_LANG_RESTORE
-                CPPFLAGS=$tmpCPPFLAGS
+                mpimajor=`grep "OMPI_MAJOR_VERSION" $MPI_INCLUDES_PATH/mpi/mpi.h | sed -e "s/#define OMPI_MAJOR_VERSION[ ]*//g"`
```

Comment:
Is this a copy/pasted version of the same test as above? Also, if I'm reading this correctly your checks are specific to OpenMPI?

Reply:
yes; currently it is specific.
I just need to look closer at different statements....

Reply:
As of MPI-1.2 there's MPI_VERSION and MPI_SUBVERSION in the standard.  Can't we just test those values to make sure they're large enough, or scream and die if we don't see them defined at all?

Reply:
> As of MPI-1.2 there's MPI_VERSION and MPI_SUBVERSION in the standard.

Those are the hooks we should use. Becomes very easy to test too!

Reply:
@BalticPinguin if I understand correctly, OpenMPI 1.6.5 claims to be MPI standard 2.1 compliant, and at least in this case, it is, since `MPI_Pack` with a *non-const pointer first argument* is in the MPI-2.1 standard, see e.g. [page 121 of this copy of the MPI-2.1 standard](https://www.mpi-forum.org/docs/mpi-2.1/mpi21-report.pdf).  So it appears that we are the ones using `MPI_Pack` incorrectly, although perhaps this API changed in MPI-2.2, I am checking that now. 

Regardless, while it will be nice to know the `MPI_VERSION` and `MPI_SUBVERSION` an implementation claims to support, we will still always need `AC_TRY_COMPILE` tests to be absolutely sure.

Reply:
The first argument of `MPI_Pack` is non-const in MPI-2.2 as well, [see pg. 121 of this PDF](https://www.mpi-forum.org/docs/mpi-2.2/mpi22-report.pdf). It first became const in [MPI-3.0](https://www.mpi-forum.org/docs/mpi-3.0/mpi30-report.pdf), so that MPI standard 2 vs. 3 has always been the issue here, not MPI 1 vs 2.